tests/test_pipe.py

```diff
@@ -0,0 +1,83 @@
+#!/usr/bin/env python
+# SPDX-FileCopyrightText: Copyright (c) 2022-2023, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+# SPDX-License-Identifier: Apache-2.0
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import gc
+import typing
+
+import pytest
+
+from morpheus.config import Config
+from morpheus.pipeline import LinearPipeline
+from morpheus.stages.input.in_memory_source_stage import InMemorySourceStage
+from morpheus.stages.output.in_memory_sink_stage import InMemorySinkStage
+from morpheus.utils.type_aliases import DataFrameType
+
+
+class SourceTestStage(InMemorySourceStage):
+
+    def __init__(self,
+                 config,
+                 dataframes: typing.List[DataFrameType],
+                 state_dict: dict,
+                 repeat: int = 1,
+                 state_key: str = "source"):
+        super().__init__(config, dataframes, repeat)
+        self._state_dict = state_dict
+        self._state_key = state_key
+
+    @property
+    def name(self) -> str:
+        return "test-source"
+
+    def __del__(self):
+        self._state_dict[self._state_key] = True
+        self._state_dict = None
+
+
+class SinkTestStage(InMemorySinkStage):
+
+    def __init__(self, config, state_dict: dict, state_key: str = "sink"):
+        super().__init__(config)
+        self._state_dict = state_dict
+        self._state_key = state_key
+
+    @property
+    def name(self) -> str:
+        return "test-sink"
+
+    def __del__(self):
+        self._state_dict[self._state_key] = True
+        self._state_dict = None
```

Comment:
The way these classes are organized could easily break if the `state_dict` is accidentally copied instead of referenced. I think a better organization would be to take a lambda function which gets executed in `__del__`. This way, the lambda function can use `nonlocal` to ensure we dont copy the `state_dict`. Something like:

```python
state_dict = {"source": False, "sink": False}

def update_state_dict(key:str):
   nonlocal state_dict
   state_dict[key] = True

pipe.set_source(SourceTestStage(config, [filter_probs_df], lambda: update_state_dict("source"))
pipe.set_source(SinkTestStage(config, lambda: update_state_dict("sink"))
```